File: src/fb2site/group_parser.py
import json
from datetime import datetime
from pathlib import Path

from .models import Post
from .language import detect_language

def parse_posts(json_file: Path) -> list[Post]:
    with json_file.open(encoding="utf-8") as f:
        data = json.load(f)

    print(f"Loaded {len(data)} posts.")

    posts = []

    for raw in data:
        body = raw.get("text", "")

        post = Post(
            id=str(raw.get("legacyId")),
            timestamp=datetime.fromisoformat(raw["time"].replace("Z", "+00:00")),
            title="",
            body=body,
            author=raw.get("user", {}).get("name"),
            facebook_url=raw.get("url"),
            group=raw.get("groupTitle"),
            language=detect_language(body),
        )

        posts.append(post)

    return posts

File: app/src/main/kotlin/in/sartaj/fb2site/group/GroupParser.kt
package `in`.sartaj.fb2site.groups

import com.fasterxml.jackson.databind.JsonNode
import `in`.sartaj.fb2site.model.NormalisedPost
import `in`.sartaj.fb2site.util.detectLanguage
import `in`.sartaj.fb2site.util.fixMojibake
import `in`.sartaj.fb2site.util.linkifyMarkdown
import java.nio.file.Path
import java.time.Instant


fun parseGroupPost(raw: JsonNode): NormalisedPost {
    val body = fixMojibake(
        raw.path("text").asText("")
	)

    return NormalisedPost(
        id = raw.path("id").asText(),
        timestamp = Instant.parse(
            raw.path("time").asText()
        ),
        title = "",
        body = body,
        links = emptyList(),
        photos = emptyList(),
        videos = emptyList(),
        language = detectLanguage(body),
        author = raw.path("user")
            .path("name")
            .asText(null),
        facebookUrl = raw.path("url")
            .asText(null),
        group = raw.path("groupTitle")
            .asText(null)
    )
}

fun parseGroupPosts(
    file: Path,
): List<NormalisedPost> {

    val json =
        readGroupPostsFile(file)

    return json.map(::parseGroupPost)
}
